jwst/outlier_detection/utils.py

```diff
@@ -20,6 +22,67 @@
 _ONE_MB = 1 << 20
 
 
+class DiskAppendableArray:
```

Comment:
Does this have unit tests? If not, would you add some to:
- demonstrate usage
- exercise the error conditions

Reply:
can do, was already starting to work on this

Reply:
Let me know if [this test suite](https://github.com/emolter/jwst/blob/JP-3741/jwst/outlier_detection/tests/test_utils.py) addresses this adequately

Reply:
Thanks! Looking at the API for this class. Can we simplify it to just what's needed for the only use in `OnDiskTimewiseOperation`? I think this would minimally involve:
- make tempdir required
- make dtype required

That way we don't need tests that make sure those options work as expected for both default and non-default conditions.

Also, can you think of any edge cases not covered by the tests? Some things that come to mind are:
- providing an invalid shape
- providing an invalid dtype
- providing an invalid tempdir

Finally, I think given the usage here having explicit cleanup of the temporary files is possible and would allow the code to not rely on the garbage collector to delete files. The same applies to the median calculator.

Reply:
Thanks for pointing this out, I will work on expanding the unit tests to cover edge cases.

Explicit garbage collection isn't something I've worried about before, so this should be a good learning experience.  Would you simply delete the filename (or temp directory) at the end of the `read` method?

Also, can you explain why allowing the garbage collector to handle it is bad?  Reading the docs of `NamedTemporaryFile`, with the default options the file should be deleted as soon as it is closed.  So is it a question of ensuring that it gets closed?  (it's not obvious to me whether or not `np.fromfile` itself closes the file)

Reply:
I added a few comments below related to this. The current code creates a tempfile (which can be used like a file) and then separately interacts with the filename (opening the file a second time and relying on synchronization via the disk which isn't always guaranteed). The comments use the tempfile directly (which requires a call to seek prior to `np.fromfile`). The comments do not re-seek the file to the end after `read`. We could re-seek to the end although in the uses here the file is written many times and only read once. That means it would be ok to make `read` also close the tempfile (by calling `self._temp_file.close()` which would also explicitly delete the file).

One downside to the above approach is that each file is kept open (as is also the case in this PR without the comments). Most systems have a limit to the number of open files (like 1024) so for a large association we may want to consider closing and opening the file for every append. We would then have to make sure the file is deleted ourselves. One way to handle this would be to just use the tempdirectory created for the median and rely on it's unlinking to delete the contained files (this is similar to how `ModelLibrary` works although here we it's easier to see where we can explicitly delete the directory).

Yeah, in this case if we can add an explicit `close` that will also delete the file. Without it we rely on the garbage collector collecting `_temp_file` which will then internally call `close`. If we can avoid relying on the garbage collector we'll have more consistent behavior.

Reply:
Thanks, that's all really helpful.

I just checked my machine, and the "soft limit" is 256 files.  Not sure the difference between the soft and hard limit (which is set to unlimited) but that might indicate that indeed it's worth opening and closing the files each time - I hope that doesn't cause too much of a performance hit, but I doubt it would.

Reply:
Files are opened and then closed when written to now; let me know if this resolves this discussion

Reply:
Thanks for the reminder and the updates. I'll mark this as resolved. It might be worth seeing if an easy test can be constructed for this (to test things are working and aid in maintenance). psutil (which jwst already requires) can get the number of open files. Here's an example test using it in stdatamodels: https://github.com/spacetelescope/stdatamodels/blob/234413f1209c50b9b103239279c813fa678dec81/tests/test_resources.py#L19